crates/iceberg-catalog/src/service/endpoint_statistics.rs

```diff
@@ -0,0 +1,291 @@
+    api::endpoints::Endpoints, request_metadata::RequestMetadata, ProjectId, WarehouseIdent,
+};
+
+/// Middleware for tracking endpoint statistics.
+///
+/// This middleware forwards information about the called endpoint to the receiver of
+/// `EndpointStatisticsTrackerTx`.
+pub(crate) async fn endpoint_statistics_middleware_fn(
+    State(tracker): State<EndpointStatisticsTrackerTx>,
+    Path(path_params): Path<HashMap<String, String>>,
+    Query(query_params): Query<HashMap<String, String>>,
+    request: Request,
+    next: Next,
+) -> Response {
+    let request_metadata = request.extensions().get::<RequestMetadata>().cloned();
+
+    let response = next.run(request).await;
+
+    if let Some(request_metadata) = request_metadata {
+        if let Err(e) = tracker
+            .0
+            .send(EndpointStatisticsMessage::EndpointCalled {
+                request_metadata,
+                response_status: response.status(),
+                path_params,
+                query_params,
+            })
+            .await
+        {
+            tracing::error!("Failed to send endpoint statistics message: {}", e);
+        };
+    } else {
+        tracing::error!(?path_params, "No request metadata found.");
+    }
+
+    response
+}
+
+/// Sender for the endpoint statistics tracker.
+#[derive(Debug, Clone)]
+pub struct EndpointStatisticsTrackerTx(tokio::sync::mpsc::Sender<EndpointStatisticsMessage>);
+
+impl EndpointStatisticsTrackerTx {
+    #[must_use]
+    pub fn new(tx: tokio::sync::mpsc::Sender<EndpointStatisticsMessage>) -> Self {
+        Self(tx)
+    }
+
+    /// Send a message to the endpoint statistics tracker.
+    ///
+    /// # Errors
+    /// If the receiver has been dropped.
+    pub async fn send(
+        &self,
+        msg: EndpointStatisticsMessage,
+    ) -> Result<(), tokio::sync::mpsc::error::SendError<EndpointStatisticsMessage>> {
+        self.0.send(msg).await
+    }
+}
+
+#[derive(Debug)]
+pub enum EndpointStatisticsMessage {
+    EndpointCalled {
+        request_metadata: RequestMetadata,
+        response_status: StatusCode,
+        path_params: HashMap<String, String>,
+        query_params: HashMap<String, String>,
+    },
+    Flush,
+    Shutdown,
+}
+
+#[derive(Debug, Default)]
+pub struct ProjectStatistics {
+    stats: HashMap<EndpointIdentifier, AtomicI64>,
+}
+
+impl ProjectStatistics {
+    #[must_use]
+    pub fn into_consumable(self) -> HashMap<EndpointIdentifier, i64> {
+        self.stats
+            .into_iter()
+            .map(|(k, v)| (k, v.load(std::sync::atomic::Ordering::Relaxed)))
+            .collect()
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct EndpointIdentifier {
+    pub uri: Endpoints,
+    pub status_code: StatusCode,
+    pub warehouse: Option<WarehouseIdent>,
+    // probably only relevant for config calls
+    pub warehouse_name: Option<String>,
+}
+
+#[derive(Debug, PartialEq, Copy, Clone)]
+pub enum FlushMode {
+    Automatic,
+    Manual,
```

Comment:
We use manual only in tests.
Maybe worth cfg(test) it?
Would need to change a match arm below as well.